Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -0,0 +1,149 @@
+/-
+Copyright (c) 2024 Judith Ludwig, Florent Schaffhauser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Judith Ludwig, Florent Schaffhauser, Yunzhou Xie, Jujian Zhang
+-/
+
+import Mathlib.RingTheory.Flat.Basic
+
+/-!
+# Faithfully flat modules
+
+A module `M` over a commutative ring `R` is *faithfully flat* if it is flat and `IM ≠ M` whenever
+`I` is a maximal ideal of `R`.
+
+## Main declaration
+
+- `Module.FaithfullyFlat`: the predicate asserting that an `R`-module `M` is faithfully flat.
+
+## Main theorems
+
+- `Module.FaithfullyFlat.iff_flat_and_proper_ideal`: an `R`-module `M` is faithfully flat iff it is
+  flat and for all proper ideals `I` of `R`, `I • M ≠ M`.
+- `Module.FaithfullyFlat.iff_flat_and_rTensor_faithful`: an `R`-module `M` is faithfully flat iff it
+  is flat and tensoring with `M` is faithful, i.e. `N ≠ 0` implies `N ⊗ M ≠ 0`.
+- `Module.FaithfullyFlat.iff_flat_and_lTensor_faithful`: an `R`-module `M` is faithfully flat iff it
+  is flat and tensoring with `M` is faithful, i.e. `N ≠ 0` implies `M ⊗ N ≠ 0`.
+
+- `Module.FaithfullyFlat.self`: the `R`-module `R` is faithfully flat.
+
+-/
+
+universe u v
+
+open TensorProduct
+
+namespace Module
+
+variable (R : Type u) (M : Type v) [CommRing R] [AddCommGroup M] [Module R M]
+
+/--
+A module `M` over a commutative ring `R` is *faithfully flat* if it is flat and,
+for all `R`-module homomorphism `f : N → N'` such that `id ⊗ f = 0`, we have `f = 0`.
+-/
+@[mk_iff] class FaithfullyFlat : Prop where
```

Comment:
I think maybe `FaithfullyFlat` can be defined by extending `Module.Flat`.

Reply:
Yes! good idea